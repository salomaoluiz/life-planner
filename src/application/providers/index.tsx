import React from "react";

import { UserProvider } from "./user";

interface Props {
  children: React.ReactNode;
}

function ApplicationProviders(props: Props) {
  return <UserProvider>{props.children}</UserProvider>;
}

export default ApplicationProviders;
