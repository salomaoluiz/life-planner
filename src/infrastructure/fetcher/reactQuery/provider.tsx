import { QueryClient, QueryClientProvider } from "@tanstack/react-query";
import React from "react";

const queryClient = new QueryClient();

interface Props {
  children: React.ReactNode;
}

function ReactQueryProvider({ children }: Props) {
  return (
    <QueryClientProvider client={queryClient}>{children}</QueryClientProvider>
  );
}

export default ReactQueryProvider;
