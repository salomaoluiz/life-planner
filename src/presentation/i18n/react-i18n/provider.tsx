import React from "react";
import i18next from "i18next";

import { I18nextProvider } from "react-i18next";
import { useI18NInitializer } from "./hooks";

interface Props {
  children: React.ReactNode;
}

export function ReactI18NPProvider({ children }: Props) {
  useI18NInitializer();
  return <I18nextProvider i18n={i18next}>{children}</I18nextProvider>;
}
