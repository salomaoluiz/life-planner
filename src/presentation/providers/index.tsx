import { I18NProvider } from "@presentation/i18n";

import React from "react";
import { ThemeProvider } from "@presentation/theme";

interface Props {
  children: React.ReactNode;
}

function PresentationProviders({ children }: Props) {
  return (
    <ThemeProvider>
      <I18NProvider>{children}</I18NProvider>
    </ThemeProvider>
  );
}

export default PresentationProviders;
