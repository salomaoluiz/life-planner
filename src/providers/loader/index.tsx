import React, { createContext, useContext, useMemo, useState } from "react";

export const loaderContexts = {
  i18n: true,
  theme: true,
};

interface LoaderContextData {
  isLoading: boolean;
  setIsLoading: (
    isLoading: boolean,
    context: keyof typeof loaderContexts,
  ) => void;
}

const LoaderContext = createContext<LoaderContextData>({} as LoaderContextData);

interface Props {
  children: React.ReactNode;
}
export const LoaderProvider = ({ children }: Props) => {
  const [isContextLoading, setContextIsLoading] = useState(loaderContexts);

  const isLoading = useMemo(() => {
    return Object.values(isContextLoading).some((value) => value);
  }, [isContextLoading]);

  const setIsLoading = (
    isLoading: boolean,
    context: keyof typeof loaderContexts,
  ) => {
    if (!loaderContexts[context]) {
      throw new Error(`Context ${context} not found`);
    }

    setContextIsLoading((prevState) => ({
      ...prevState,
      [context]: isLoading,
    }));
  };

  return (
    <LoaderContext.Provider value={{ isLoading, setIsLoading }}>
      {children}
    </LoaderContext.Provider>
  );
};

export function useProviderLoader() {
  const context = useContext(LoaderContext);

  if (Object.keys(context).length === 0) {
    throw new Error("useLoader must be used within an LoaderProvider");
  }

  return context;
}
