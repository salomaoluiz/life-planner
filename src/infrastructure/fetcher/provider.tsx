import React from "react";
import { ReactQueryProvider } from "@infrastructure/fetcher/reactQuery";

interface Props {
  children: React.ReactNode;
}

function FetcherProvider({ children }: Props) {
  return <ReactQueryProvider>{children}</ReactQueryProvider>;
}

export default FetcherProvider;
