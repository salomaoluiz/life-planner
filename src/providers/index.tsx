import React from "react";
import { LoaderProvider } from "./loader";

import PresentationProviders from "@presentation/providers";

interface Props {
  children: React.ReactNode;
}

function GlobalProviders({ children }: Props) {
  return (
    <LoaderProvider>
      <PresentationProviders>{children}</PresentationProviders>
    </LoaderProvider>
  );
}

export default GlobalProviders;
