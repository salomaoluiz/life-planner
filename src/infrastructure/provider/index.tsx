import React from "react";

import { FetcherProvider } from "@infrastructure/fetcher";
import * as GoogleOAuth from "@infrastructure/googleOAuth";

interface Props {
  children: React.ReactNode;
}
function InfrastructureProvider({ children }: Props) {
  GoogleOAuth.initialize();
  return <FetcherProvider>{children}</FetcherProvider>;
}

export default InfrastructureProvider;
