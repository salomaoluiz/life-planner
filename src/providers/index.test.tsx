import { View } from "react-native";

import { render, screen } from "@tests";

import * as applicationProviders from "@application/providers";
import * as infrastructureProvider from "@infrastructure/provider";
import * as presentationProviders from "@presentation/providers";

import GlobalProviders from "./";
import * as loaderProvider from "./loader";

jest
  .spyOn(presentationProviders, "default")
  .mockImplementation(({ children }) => (
    <View testID={"presentation-providers"}>{children}</View>
  ));

jest
  .spyOn(infrastructureProvider, "default")
  .mockImplementation(({ children }) => (
    <View testID={"infrastructure-providers"}>{children}</View>
  ));

jest
  .spyOn(applicationProviders, "default")
  .mockImplementation(({ children }) => (
    <View testID={"application-providers"}>{children}</View>
  ));

jest
  .spyOn(loaderProvider, "LoaderProvider")
  .mockImplementation(({ children }) => (
    <View testID={"loader-provider"}>{children}</View>
  ));

function Children() {
  return <View testID={"children-element"} />;
}

function setup() {
  render(<Children />, { wrapper: GlobalProviders });
}

it('SHOULD render "PresentationProviders", "ApplicationProviders" and "LoaderProvider"', () => {
  setup();

  expect(screen.getByTestId("presentation-providers")).toBeOnTheScreen();
  expect(screen.getByTestId("application-providers")).toBeOnTheScreen();
  expect(screen.getByTestId("infrastructure-providers")).toBeOnTheScreen();
  expect(screen.getByTestId("loader-provider")).toBeOnTheScreen();
  expect(screen.getByTestId("children-element")).toBeOnTheScreen();
  expect(screen.toJSON()).toMatchSnapshot();
});
