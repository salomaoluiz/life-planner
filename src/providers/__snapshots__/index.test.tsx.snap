// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SHOULD render "PresentationProviders", "ApplicationProviders" and "LoaderProvider" 1`] = `
<View
  testID="loader-provider"
>
  <View
    testID="infrastructure-providers"
  >
    <View
      testID="application-providers"
    >
      <View
        testID="presentation-providers"
      >
        <View
          testID="children-element"
        />
      </View>
    </View>
  </View>
</View>
`;
